poethepoet/task/shell.py

```diff
@@ -31,42 +41,104 @@ def _handle_run(
-        shell_from_env = shutil.which(os.environ.get("SHELL", "bash"))
-        if shell_from_env:
-            return [shell_from_env]
-
-        # Try locate a bash from the environment
-        bash_from_env = shutil.which("bash")
-        if bash_from_env:
-            return [bash_from_env]
-
-        # Or check specifically for git bash
-        bash_from_git = shutil.which("C:\\Program Files\\Git\\bin\\bash.exe")
-        if bash_from_git:
-            return [bash_from_git]
-
-        # or use bash from wsl if it's available
-        wsl = shutil.which("wsl")
-        if wsl and subprocess.run(["wsl", "bash"], capture_output=True).returncode > 0:
-            return [wsl, "bash"]
-
-        # Fail: if there is a bash out there, we don't know how to get to it
-        # > We don't know how to access wsl bash from python installed from python.org
-        raise PoeException(
-            "Couldn't locate bash executable to run shell task. Installing WSL should "
-            "fix this."
+    def _get_interpreter_config(self) -> Tuple[str, ...]:
+        result: Union[str, Tuple[str, ...]] = self.options.get(
+            "interpreter", self._config.shell_interpreter
         )
+        if isinstance(result, str):
+            return (result,)
+        return tuple(result)
+
+    def resolve_interpreter(self) -> Optional[str]:
+        for item in self._get_interpreter_config():
+            return self._locate_interpreter(item)
+        return None
+
+    def _locate_interpreter(self, interpreter: str) -> Optional[str]:
+        result = None
+
+        # Try use $SHELL from the environment as a hint
+        shell_var = environ.get("SHELL", "")
+        if shell_var.endswith(f"/{interpreter}") and which(shell_var) == shell_var:
+            result = shell_var
+
+        elif interpreter == "posix":
+            # look for any known posix shell
+            result = (
+                self._locate_interpreter("sh")
+                or self._locate_interpreter("bash")
+                or self._locate_interpreter("zsh")
+            )
+
+        elif interpreter == "sh":
+            result = which("sh") or which("/bin/sh")
+
+        elif interpreter == "bash":
+            result = which("bash") or which("/bin/bash")
+
+            # Specifically look for git bash on windows
+            if result is None and self._is_windows:
+                result = which("C:\\Program Files\\Git\\bin\\bash.exe")
+
+        elif interpreter == "zsh":
+            result = which("zsh") or which("/bin/zsh")
+
+        elif interpreter == "fish":
+            result = which("fish") or which("/bin/fish")
+
+        elif interpreter == "pwsh":
+            result = which("pwsh") or which("powershell")
+
+            # Specifically look in a known location on windows
+            if result is None and self._is_windows:
+                result = which(
+                    "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.EXE"
+                )
```

Comment:
I would suggest doing this only when using `powershell`, not `pwsh`, as the two are not fully cross-compatible.